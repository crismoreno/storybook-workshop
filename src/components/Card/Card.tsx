import React from 'react';
import { ReactNode } from 'react';
import styled from 'styled-components';
import '../../styles';
import { colorsLib, fontsLib } from '../../styles';

export type CardProps = {
  image: string;
  onClick?: () => void;
  title: ReactNode | string;
  subtitle?: string;
  makeHoverEffect?: boolean | false;
};

type StyledCardProps = {
  $hasOnClick?: boolean;
  $isTitleString?: boolean;
  $makeHoverEffect?: boolean;
};

const Card = ({
  image,
  onClick,
  title,
  subtitle,
  makeHoverEffect,
}: CardProps): JSX.Element => {
  const isTitleString = typeof title === 'string';
  const mustTitleBeTruncated = subtitle && isTitleString && title.length >= 30;

  return (
    <StyledCard
      onClick={onClick}
      $hasOnClick={Boolean(onClick)}
      $isTitleString={isTitleString}
      $makeHoverEffect={makeHoverEffect}
    >
      <div className="card-content flex flex-col">
        <div className="img-container">
          <img className="project-image" src={image} />
        </div>
        <div className="project-details flex flex-col justify-center p-3">
          <span className="title" aria-label="title">
            {mustTitleBeTruncated ? title.substring(0, 30) + '\u2026' : title}
          </span>
          <p className="subtitle">{subtitle}</p>
        </div>
      </div>
    </StyledCard>
  );
};

const StyledCard = styled.div<StyledCardProps>`
  box-shadow:
    0 15px 35px rgba(50, 50, 93, 0.1),
    0 5px 15px rgba(0, 0, 0, 0.07);
  border-radius: 7px;
  height: fit-content;
  border: none;
  width: 265px;

  .img-container {
    border-top-left-radius: 7px;
    border-top-right-radius: 7px;
    overflow: hidden;
    img {
      border-top-left-radius: 5px;
      border-top-right-radius: 5px;
      transition: transform 1s ease;
    }
  }

  &:hover {
    .img-container {
      ${({ $makeHoverEffect }) =>
        $makeHoverEffect &&
        ` border-top-left-radius: 5px;
					border-top-right-radius: 5px;
				`}

      img {
        ${({ $makeHoverEffect }) =>
          $makeHoverEffect &&
          `	border-top-left-radius: 5px;
						border-top-right-radius: 5px;
        transform: scale(1.1);`}
      }
    }
  }

  .project-details {
    font-size: 13px;
    color: ${colorsLib.dark};
    font-family: ${fontsLib.sourceSansPro};
    .title {
      ${({ $isTitleString }) => $isTitleString && `text-transform: uppercase `}
    }
    .subtitle {
      text-transform: none;
    }
  }
  ${({ $hasOnClick }) => $hasOnClick && `cursor: pointer`}
`;

export default Card;
