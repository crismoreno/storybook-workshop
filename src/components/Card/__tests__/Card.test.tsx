import React from 'react';
import '@testing-library/jest-dom';
import { render, act } from '@testing-library/react';
import type { RenderResult } from '@testing-library/react';
import { screen } from '@testing-library/react';
import type { CardProps } from '../Card';
import { createRoot } from 'react-dom/client';

import Card from '../Card';

describe('Button', () => {
  const defaultProps: CardProps = {
    image: 'https://i.blogs.es/96c029/windows-xp-sin-soporte/1366_2000.jpg',
    title: 'Test title',
  };

  const getComponent = (props = defaultProps): JSX.Element => (
    <Card {...props}>Test Button</Card>
  );

  const setup = (props = defaultProps): RenderResult =>
    render(getComponent(props));

  it('Should render correctly', async () => {
    const div = document.createElement('div');
    act(() => {
      const root = createRoot(div);
      root.render(getComponent());
      root.unmount();
    });
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should truncate title when exceeds 30 chars and there is a subtitle', () => {
    setup({
      ...defaultProps,
      title: 'This is a super super super super super super long title',
      subtitle: 'This is a subtitle',
    });

    const text = screen.getByText('This is a super super super suâ€¦');

    expect(text).toBeInTheDocument();
  });

  it('Should not truncate title when exceeds 30 chars but there is no subtitle', () => {
    setup({
      ...defaultProps,
      title: 'This is a super super super super super super long title',
    });

    const text = screen.getByText(
      'This is a super super super super super super long title'
    );

    expect(text).toBeInTheDocument();
  });
});
