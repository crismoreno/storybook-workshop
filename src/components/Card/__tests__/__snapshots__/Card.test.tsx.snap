// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot 1`] = `
<div>
  <div
    class="sc-bdnyFh eXOPrQ"
  >
    <div
      class="card-content flex flex-col"
    >
      <div
        class="img-container"
      >
        <img
          class="project-image"
          src="https://i.blogs.es/96c029/windows-xp-sin-soporte/1366_2000.jpg"
        />
      </div>
      <div
        class="project-details flex flex-col justify-center p-3"
      >
        <span
          aria-label="title"
          class="title"
        >
          Test title
        </span>
        <p
          class="subtitle"
        />
      </div>
    </div>
  </div>
</div>
`;
